pkg/kp/consulutil/watch.go

```diff
@@ -35,35 +38,58 @@ func WatchPrefix(
 	if pause < 250*time.Millisecond {
 		pause = 250 * time.Millisecond
 	}
+
+	listLatencyHistogram, consulLatencyHistogram, outputPairsBlocking, outputPairsHistogram := watchHistograms("prefix")
+
+	var (
+		safeListStart    time.Time
+		outputPairsStart time.Time
+	)
 	for {
 		select {
 		case <-done:
 			return
 		case <-timer.C:
 		}
-		timer.Reset(pause) // upper bound on request rate
+		safeListStart = time.Now()
 		pairs, queryMeta, err := SafeList(clientKV, done, prefix, &api.QueryOptions{
-			WaitIndex: currentIndex,
+			WaitIndex:  currentIndex,
+			AllowStale: true,
 		})
+		listLatencyHistogram.Update(int64(time.Since(safeListStart) / time.Millisecond))
+		timer.Reset(pause)
 		switch err {
 		case CanceledError:
 			return
 		case nil:
 			currentIndex = queryMeta.LastIndex
+			consulLatencyHistogram.Update(int64(queryMeta.RequestTime))
+			outputPairsStart = time.Now()
 			select {
 			case <-done:
 			case outPairs <- pairs:
 			}
+			outputPairsBlocking.Update(int64(time.Since(outputPairsStart) / time.Millisecond))
+			outputPairsHistogram.Update(int64(sizeInBytes(pairs)))
 		default:
 			select {
 			case <-done:
 			case outErrors <- err:
+				timer.Reset(2*time.Second + pause)
```

Comment:
This code went through a lot of iterations locally because it arrived here in review.